service/src/main/java/dk/dbc/rr/oai/OaiBean.java

```diff
@@ -40,14 +50,66 @@
     private static final Logger log = LoggerFactory.getLogger(OaiBean.class);
 
     @Inject
-    IndexHtml indexHtml;
+    public Config config;
+
+    @Inject
+    public RemoteIp remoteIp;
+
+    @Inject
+    public ForsRights forsRights;
+
+    @Inject
+    public IndexHtml indexHtml;
 
     @GET
-    public Response oai(@Context UriInfo uriInfo) {
-        log.info("index.html");
+    public Response oai(@Context UriInfo uriInfo,
+                        @Context HttpServletRequest httpRequest,
+                        @Context HttpHeaders headers) {
+
+        MultivaluedMap<String, String> params = uriInfo.getQueryParameters();
+        if (params.isEmpty()) {
+            log.info("index.html");
+            return Response.ok()
+                    .type(MediaType.TEXT_HTML_TYPE)
+                    .entity(indexHtml.getInputStream())
+                    .build();
+        }
+
+        String triple = params.getFirst("identity");
+        if (triple == null || triple.split(":", 3).length != 3)
+            triple = headers.getHeaderString("Identity");
+        String clientIp = remoteIp.clientIp(httpRequest.getRemoteAddr(),
+                                            headers.getHeaderString("X-Forwarded-For"));
+        log.trace("tripplet = {}; clientIp = {}", triple, clientIp);
+
         return Response.ok()
-                .type(MediaType.TEXT_HTML_TYPE)
-                .entity(indexHtml.getInputStream())
+                .type(MediaType.APPLICATION_XML_TYPE)
+                .entity(processOaiRequest(triple, clientIp, params))
                 .build();
     }
+
+    byte[] processOaiRequest(String triple, String clientIp, MultivaluedMap<String, String> params) {
+        Set<String> allowedSets = getAllowedSets(triple, clientIp);
+
+        if (allowedSets.isEmpty())
+            throw new ServerErrorException(UNAUTHORIZED);
+
+        throw new ServerErrorException("Not implemented yet!", INTERNAL_SERVER_ERROR);
+    }
+
+    Set<String> getAllowedSets(String triple, String clientIp) throws ServerErrorException {
+        Set<String> allowedSets = EMPTY_SET;
+        if (config.isAuthenticationDisabled()) {
+            allowedSets = config.getAllForsRightsSets();
+        } else {
+            try {
+                allowedSets = forsRights.authorized(triple, clientIp);
+            } catch (IOException ex) {
+                log.error("Error validating user: {}", ex.getMessage());
```

Comment:
"validating user"? Not saying it's not what is happening here, but I think the message here could be clearer.

Reply:
This is only if a user validation has technical problems ie. the server is unreachable. If the call is successful but the user isn't validated allowed sets will be empty. Given the design of ForsRights, we cannot tell if the the user didn't validate (password) or the rights to this user wasn't granted (permission).

Reply:
Error communicating with ForsRights